<launch>

   <!-- Rviz Visualization -->
   <node pkg="rviz" name="rviz" type="rviz" args="-d $(find ouster_ros)/rviz/config.rviz"/>

   <arg name="sensor_hostname" default="10.5.0.78" doc="hostname or IP in dotted decimal form of the sensor"/>
   <arg name="udp_dest"        default="10.5.0.1" doc="hostname or IP where the sensor will send data packets"/>
   <arg name="lidar_port"      default="7502" doc="port to which the sensor should send lidar data"/>
   <arg name="imu_port"        default="7503" doc="port to which the sensor should send imu data"/>
   <arg name="lidar_mode"      default="1024x10" doc="resolution and rate: either 512x10, 512x20, 1024x10, 1024x20, or 2048x10"/>
   <arg name="timestamp_mode"  default="TIME_FROM_INTERNAL_OSC" doc="method used to timestamp measurements: TIME_FROM_INTERNAL_OSC, TIME_FROM_SYNC_PULSE_IN, TIME_FROM_PTP_1588"/>
   <arg name="tf_prefix"       default="" doc="namespace for tf transforms"/>
   <arg name="tf_inverse"      default="true" doc="inverse tf os_sensor - os_lidar"/>
   <arg name="stamp_offset"    default="0.0" doc="timestamp (in secs) to be added"/>
   <arg name="max_sync_diff"   default="1.5" doc="max sync diff before publishing"/>
   <arg name="enable_img_node" default="false"/>
   <arg name="enab_split_node" default="true"/>
   <arg name="metadata"        default=""/>

   <arg name="up_mirror_points_x"     default ="[-0.04, 0.04, 0.04, -0.04]"/>
   <arg name="up_mirror_points_y"     default ="[-0.051, -0.051, -0.1217, -0.1217]"/>
   <arg name="up_mirror_points_z"     default ="[0.0, 0.0, 0.0842, 0.0842]"/>

   <arg name="down_mirror_points_x"   default ="[-0.04, 0.04, 0.04, -0.04]"/>
   <arg name="down_mirror_points_y"   default ="[-0.051, -0.051, -0.1217, -0.1217]"/>
   <arg name="down_mirror_points_z"   default ="[0.0, 0.0, -0.0842, -0.0842]"/>

   <node pkg="ouster_ros" name="os_node" type="os_node" output="screen" required="true">
      <param name="~/lidar_mode"     type="string" value="$(arg lidar_mode)"/>
      <param name="~/timestamp_mode" type="string" value="$(arg timestamp_mode)"/>
      <param name="~/sensor_hostname" value="$(arg sensor_hostname)"/>
      <param name="~/udp_dest"        value="$(arg udp_dest)"/>
      <param name="~/lidar_port"      value="$(arg lidar_port)"/>
      <param name="~/imu_port"        value="$(arg imu_port)"/>
      <param name="~/tf_inverse"      value="$(arg tf_inverse)"/>
      <param name="~/stamp_offset"    value="$(arg stamp_offset)"/>
      <param name="~/max_sync_diff"   value="$(arg max_sync_diff)"/>

      <rosparam param="/up_mirror_points_x"   subst_value="True">$(arg up_mirror_points_x)</rosparam>
      <rosparam param="/up_mirror_points_y"   subst_value="True">$(arg up_mirror_points_y)</rosparam>
      <rosparam param="/up_mirror_points_z"   subst_value="True">$(arg up_mirror_points_z)</rosparam>

      <rosparam param="/down_mirror_points_x" subst_value="True">$(arg down_mirror_points_x)</rosparam>
      <rosparam param="/down_mirror_points_y" subst_value="True">$(arg down_mirror_points_y)</rosparam>
      <rosparam param="/down_mirror_points_z" subst_value="True">$(arg down_mirror_points_z)</rosparam>


   </node>

   <group if="$(arg enable_img_node)">
      <node pkg="ouster_ros" name="os_img_node" type="os_img_node" output="screen" required="false">
         <param name="~/metadata" value="$(arg metadata)"/>
      </node>
   </group>

   <group if="$(arg enab_split_node)">
      <node pkg="ouster_ros" name="os_split_cloud_node" type="os_split_cloud_node" output="screen" required="false">
      
         <rosparam param="/up_mirror_points_x"   subst_value="True">$(arg up_mirror_points_x)</rosparam>
         <rosparam param="/up_mirror_points_y"   subst_value="True">$(arg up_mirror_points_y)</rosparam>
         <rosparam param="/up_mirror_points_z"   subst_value="True">$(arg up_mirror_points_z)</rosparam>
         <rosparam param="/down_mirror_points_x" subst_value="True">$(arg down_mirror_points_x)</rosparam>
         <rosparam param="/down_mirror_points_y" subst_value="True">$(arg down_mirror_points_y)</rosparam>
         <rosparam param="/down_mirror_points_z" subst_value="True">$(arg down_mirror_points_z)</rosparam>

      </node>
   </group>

</launch>
